nautobot/dcim/views.py

```diff
@@ -1609,6 +1953,364 @@ class DeviceBulkDeleteView(generic.BulkDeleteView):
+                            new_components
+                        ):
+                            raise ObjectDoesNotExist
+
+                except IntegrityError:
+                    pass
+
+                except ObjectDoesNotExist:
+                    msg = "Component creation failed due to object-level permissions violation"
+                    self.logger.debug(msg)
+                    form.add_error(None, msg)
+
+                if not form.errors:
+                    msg = f"Added {len(new_components)} {model_name} to {len(form.cleaned_data['pk'])} {parent_model_name}."
+                    self.logger.info(msg)
+                    messages.success(request, msg)
+
+                    return redirect(self.get_return_url(request))
+
+            else:
+                self.logger.debug("Form validation failed: %s", form.errors)
+
+        else:
+            form = bulk_component_form(model, initial={"pk": pk_list})
+
+        return Response(
+            {
+                "template": "generic/object_bulk_add_component.html",
+                "form": form,
+                "parent_model_name": parent_model_name,
+                "model_name": model_name,
+                "table": table,
+                "return_url": self.get_return_url(request),
+            },
+        )
+
+
+class ModuleUIViewSet(BulkComponentCreateUIViewSetMixin, NautobotUIViewSet):
+    queryset = Module.objects.all()
+    filterset_class = filters.ModuleFilterSet
+    filterset_form_class = forms.ModuleFilterForm
+    form_class = forms.ModuleForm
+    bulk_update_form_class = forms.ModuleBulkEditForm
+    serializer_class = serializers.ModuleSerializer
+    table_class = tables.ModuleTable
+
+    def get_extra_context(self, request, instance):
+        context = super().get_extra_context(request, instance)
+        if instance:
+            context["modulebay_count"] = instance.module_bays.count()
+            populated_module_count = instance.module_bays.filter(installed_module__isnull=False).count()
+            context["module_count"] = f"{populated_module_count}/{context['modulebay_count']}"
+        if self.action in ["create", "update"]:
+            context["active_parent_tab"] = self._get_edit_view_active_parent_tab(request)
+        return context
+
+    def _get_edit_view_active_parent_tab(self, request):
+        active_parent_tab = "device"
+        form_class = self.get_form_class()
+        form = form_class(
+            data=request.POST,
+            files=request.FILES,
+            initial=normalize_querydict(request.GET, form_class=form_class),
+            instance=self.get_object(),
+        )
+        if form["parent_module_bay_module"].initial:
+            active_parent_tab = "module"
+        elif form["location"].initial:
+            active_parent_tab = "location"
+
+        return active_parent_tab
+
+    @action(detail=True)
+    def consoleports(self, request, *args, **kwargs):
+        # TODO: permissions enforcement?
```

Comment:
Were you able to sort this out?

Reply:
On day 2 of trying to get this sorted out 😞

Reply:
Today I learned that you should always call `self.get_object()` and `self.get_queryset()` in NautobotUIViewsets if you want permissions to be enforced

Reply:
Oh yeah, that's a big footgun. :-(